### DISCUSSION TITLE
Allow the ability to throw custom errors from verify

### DISCUSSION THREAD

**Discussion by MarcDonald** [sergiodxa/remix-auth#203] [Ideas]
At the moment it seems that any error thrown in the `verify` method of a strategy gets turned into an `AuthorizationError` by the time it gets back to the loader/action. I would like to use a custom error that is thrown in verify and can be caught and handled separately from AuthorizationError.

E.g. (pseudocode)

```
verify() {

verifyUser();

if(cantFindUser) {
  throw new UserNotFoundError(providerProfile);
} else if (somethingWentWrong) {
  throw new AuthorizationError();
} else {
  return user;
}
```

```
loader() {
  try {
     await authenticator.authenticate();
  } catch(e) {
    if(e instanceOf UserNotFoundError) {
      registerNewUser(e.providerProfile);
    } else {
      redirectTo('/error');
    }
  }
}
```

If this looks like complete nonsense and is a totally wrong way of thinking about how this flow should be handled please let me know what the proper way to handle it would be, thanks in advance!

**Comment by sergiodxa** [comment]
This is now possible with the `error.cause`, any AuthorizationError instance contains a `cause` property that includes the original error

```ts
try {
  await authenticator.authenticate("strategy", request, {
    throwOnError: true,
    successRedirect: "/private"
  })
} catch (error) {
  if (error instanceof AuthorizationError) {
    if (error.cause instanceof CustomErrorClass) {
      // do something here
    }
    // or here
  }
  // here
}
```

**Reply by MarcDonald** [reply]
Amazing, thank you!
